crates/conjure_core/src/rule_engine/rule.rs

```diff
@@ -105,7 +105,6 @@ impl Reduction {
         // TODO: (yb33) Remove it when we change constraints to a vector
         if let Expression::And(_, exprs) = &self.new_top {
             if exprs.is_empty() {
-                model.constraints = self.new_expression.clone();
```

Comment:
Removing this fixes my naive rewriter, but breaks the current one

Reply:
When we had `Nothing` it used to be 

```rust
    pub fn apply(self, model: &mut Model) {
        model.variables.extend(self.symbols); // Add new assignments to the symbol table
        if let Expression::Nothing = self.new_top {
            model.constraints = self.new_expression.clone();
        } else {
            model.constraints = match self.new_expression {
                Expression::And(metadata, mut exprs) => {
                    // Avoid creating a nested conjunction
                    exprs.push(self.new_top.clone());
                    Expression::And(metadata.clone_dirty(), exprs)
                }
                _ => Expression::And(
                    Metadata::new(),
                    vec![self.new_expression.clone(), self.new_top],
                ),
            };
        }
```

If new_top is `Nothing` - we can just replace our current model.constraints with our current reduction without having to extend the model with new_top statement.

Same applies to this version of code, but representation of `Nothing` is different. If `new_top` is an empty `And`, we just update the model with our current reduction without new_top.

Reply:
(That is what I **think** logic is supposed to be)

Reply:
Thanks for taking a look; That all makes sens!


 I guess the implicit assumption here (that isn't valid in my case) is that apply adds a new expression to the top level? 

 I left a comment on your PR about this; I think we should make this only apply the side effects, and move all the  new'expression modification code to the rewriter. This would make it no longer rely on the fact that a rewriter rewrites a top level expression, which may not necessarily be true ( by using zippers, cursors, contexts and so on)